src/api/wallet/WalletApiImpl.ts

```diff
@@ -0,0 +1,254 @@
+
+// to track chain state, be that current account balance or token balance of that account
+// or any data on the chain belonging to that account
+// we need to refetch that data when
+// 1: network changes
+// 2: account changes
+// 3: new block is mined
+
+interface BlockchainUpdatePrompt {
+  account: string
+  chainId: number
+  blockHeader: BlockHeader | null
+}
+
+type BlockchainUpdatePromptCallback = (callback: (changedChainData: BlockchainUpdatePrompt) => void) => Command
+
+// provides subscription to blockhain updates for account/network/block
+const subscribeToBlockchainUpdate = ({
+  provider,
+  subscriptions,
+  web3,
+}: {
+  provider: Provider
+  subscriptions?: Subscriptions
+  web3: Web3
+}): BlockchainUpdatePromptCallback => {
+  const subs = subscriptions || createSubscriptions(provider)
+
+  let networkUpdate: (callback: (chainId: number) => void) => Command
+
+  if (isMetamaskSubscriptions(subs)) networkUpdate = (cb): Command => subs.onNetworkChanged(networkId => cb(+networkId))
+  if (isWalletConnectSubscriptions(subs)) networkUpdate = subs.onChainChanged
+
+  const accountsUpdate = subs.onAccountsChanged
+
+  const blockUpdate = (cb: (blockHeader: BlockHeader) => void): Command => {
+    const blockSub = web3.eth.subscribe('newBlockHeaders').on('data', cb)
+    return (): void => {
+      blockSub.unsubscribe()
+    }
+  }
+
+  const {
+    accounts: [account],
+    chainId,
+  } = getProviderState(provider)
+
+  let blockchainPrompt: BlockchainUpdatePrompt = {
+    account,
+    chainId: +chainId,
+    blockHeader: null,
+  }
+
+  const subscriptionHOC: BlockchainUpdatePromptCallback = callback => {
+    const unsubNetwork = networkUpdate(chainId => {
+      blockchainPrompt = { ...blockchainPrompt, chainId }
+      log('chainId changed:', chainId)
+      callback(blockchainPrompt)
+    })
+    const unsubAccounts = accountsUpdate(([account]) => {
+      blockchainPrompt = { ...blockchainPrompt, account }
+      log('accounts changed:', account)
+      callback(blockchainPrompt)
+    })
+
+    const unsubBlock = blockUpdate(blockHeader => {
+      blockchainPrompt = { ...blockchainPrompt, blockHeader }
+      log('block changed:', blockHeader.number)
+      callback(blockchainPrompt)
+    })
+
+    return (): void => {
+      unsubNetwork()
+      unsubAccounts()
+      unsubBlock()
+    }
+  }
+
+  return subscriptionHOC
+}
+
+// const AUTOCONNECT = process.env.AUTOCONNECT === 'true'
+
+const wcOptions: WalletConnectInits = {
+  package: WalletConnectProvider,
+  options: {
+    // TODO get infuraId from .env
+    infuraId: '8b4d9b4306294d2e92e0775ff1075066',
+  },
+}
+
+/**
+ * Basic implementation of Wallet API
+ */
+export class WalletApiImpl implements WalletApi {
+  private _listeners: ((walletInfo: WalletInfo) => void)[]
+  private _provider: Provider | null
+  private _web3: Web3 | null
+
+  private _unsubscribe: Command = () => {}
+
+  public constructor() {
+    this._listeners = []
```

Comment:
Should this logic be refactored into a `WalletBase` since it's copy pasted from the mock?

addOnChangeWalletInfo, _listeners, _notifyListeners, etc.

Reply:
Or maybe even a mixin https://www.typescriptlang.org/docs/handbook/mixins.html

Reply:
Yeah, sure, can even extend an EventEmitter, like web3 does